subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,695 @@
-                cannot automatically figure out whether the task is up-to-date or not. To use Gradle's up-to-date checking,
-                you need to declare the inputs and outputs of the task.
-            </para>
-            <para>Each task has an <literal>inputs</literal> and <literal>outputs</literal> property, which you use to
-                declare the inputs and outputs of the task. Below, we have changed our example to declare that it takes
-                the source XML file as an input and produces output to a destination directory. Let's run it a couple
-                of times.
+
+            <figure>
+                <title>Example task inputs and outputs</title>
+                <imageobject>
+                    <imagedata fileref="img/taskInputsOutputs.png" width="160mm"/>
+                </imageobject>
+            </figure>
+
+            <para>An important characteristic of an input is that it affects one or more outputs, as
+                you can see from the previous figure. Different bytecode is generated depending on the
+                content of the source files and the minimum version of the Java runtime you want to 
+                run the code on. That makes them task inputs. But whether the compilation runs in a
+                forked process or not, determined by the <literal>fork</literal> property, has no
+                impact on what bytecode gets generated. In Gradle terminology, <literal>fork</literal>
+                is just an internal task property.
             </para>
-            <sample id="incrementalTask" dir="userguide/tasks/incrementalBuild/inputsAndOutputs" title="Declaring the inputs and outputs of a task">
-                <sourcefile file="build.gradle"/>
-                <output args="transform" outputFile="generatorTask.out" ignoreExtraLines="true"/>
-                <output args="transform" ignoreExtraLines="true"/>
-            </sample>
-            <para>Now, Gradle knows which files to check to determine whether the task is up-to-date or not.</para>
-            <para>
-                The task's <literal>inputs</literal> property is of type <apilink class="org.gradle.api.tasks.TaskInputs"/>.
-                The task's <literal>outputs</literal> property is of type <apilink class="org.gradle.api.tasks.TaskOutputs"/>.
+            <para>As part of incremental build, Gradle tests whether any of the task inputs or outputs
+                have changed since the last build. If they haven’t, Gradle can consider the task up to
+                date and therefore skip executing its actions. Also note that incremental build won’t
+                work unless a task has at least one task output, although tasks usually have at least
+                one input as well.
             </para>
-            <para>
-                A task with no defined outputs will <emphasis>never</emphasis> be considered up-to-date.
```

Comment:
> A task with no defined outputs will *never* be considered up-to-date.

This sentence was removed from the manual last year. Is that because it's no longer true? If it is true, I would find it really helpful to keep the sentence in the manual.

My assumption was that a task with inputs but no outputs would run when the inputs change, rather than always run.  The fact that this is not the case is surprising (in fact, it seems like Gradle should warn if you define inputs but no outputs) and should still be documented if true...

Reply:
It was misleading.  For [lifecycle tasks](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:lifecycle_tasks), a task can have no inputs, outputs or actions and still report being UP-TO-DATE or not.

We tried to describe the different task outcomes more [here](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:task_outcomes).

Reply:
I think all tasks have some kind of output (calculate a property, produce some console output, populate a database), but not all outputs are understandable by Gradle.  We're limited to outputs that are files right now.  So a task with no _file_ outputs means Gradle doesn't have a way to know if the task's work is done.

A rule like "if you define inputs and no outputs, then it should re-run whenever the inputs change and not if they don't" assumes that the output of the task cannot change without a noticeable change in the inputs.  This falls down for things like "deploy this application" (which could have inputs and non-file outputs).  We're missing a feature that allows you to say some arbitrary serializable thing is the representation of the task's outputs that aren't files.

`outputs.upToDateWhen { true }` adds to Gradle's up-to-date checking, so this will still always execute the task if there are no other outputs defined.  From the [Javadoc](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskOutputs.html#upToDateWhen(org.gradle.api.specs.Spec)), task outputs are considered out-of-date if this returns false.  It doesn't mean that the task is considered up-to-date if it returns true.

Reply:
So the example that led me here is a linter.  If the input files haven't changed, I don't want to re-run the linter.  (If I really believe there's some sort of non-determinism in my linter, I can use --rerun-tasks, I guess.)  I guess this counts as "produce some console output" but I don't know how to express that other than as `outputs.upToDateWhen { true }`.

Reply:
The easiest thing to do is to put your linter's output into a report file and write to the console too (or make that configurable).  That's what the built-in Gradle code quality tasks do. 

To be clear, `outputs.upToDateWhen { true }` does not mean your outputs are up-to-date.  It only means that the additional check (on top of the check that Gradle already does) considers the task's outputs up-to-date.  If the task doesn't have any outputs declared, Gradle considers the task's outputs as unknown/out-of-date.

Please feel free to open a new topic on https://discuss.gradle.org/ if you have more questions.